File: 0438_fnd_all_anagrams_in_a_string/find_anagrams.py
from collections import Counter
from typing import List


def find_anagrams(self, s: str, p: str) -> List[int]:
    """
    Time: O(s + p)
    Space: O(s + p)
    """
    s_size, p_size = len(s), len(p)
    if p_size > s_size:
        return []
    res = []
    s_counter = Counter()
    p_counter = Counter(p)
    for i in range(s_size):
        s_counter[s[i]] += 1
        if i >= p_size:
            if s_counter[s[i - p_size]] == 1:
                del s_counter[s[i - p_size]]
            else:
                s_counter[s[i - p_size]] -= 1
        if s_counter == p_counter:
            res.append(i - p_size + 1)
    return res

File: 0438_fnd_all_anagrams_in_a_string/find_anagrams.go
package main

/**
 * Time: O(s + p)
 * Space: O(p)
 */
func findAnagrams(s string, p string) []int {
	pCount := make(map[byte]int)
	counter := len(p)
	result := make([]int, 0)
	for i := 0; i < len(p); i++ {
		pCount[p[i]]++
	}
	begin, end := 0, 0
	for end < len(s) {
		val := pCount[s[end]]
		pCount[s[end]]--
		if val > 0 {
			counter--
		}
		end++
		if counter == 0 {
			result = append(result, begin)
		}
		if end >= len(p) {
			val := pCount[s[begin]]
			pCount[s[begin]]++
			if val >= 0 {
				counter++
			}
			begin++
		}
	}
	return result
}

func main() {

}
